### PR TITLE
Remap OpenZeppelin contract and create a Library for counter

### PR DESCRIPTION
**Pull request by olasunkanmi-SE**
* Transform Counter contract into a reusable library with structured Counter type
* Add OpenZeppelin contract remappings in foundry.toml
* Introduce a new contract ERC20TokenManager that inherits from ERC20 and implements AccessControl
* Define MINTER_ROLE and grant it to the contract deployer
* Implement mint and burn functions to handle token creation and destruction
* Use OpenZeppelin contracts as dependencies (ERC20, AccessControl)
- Introduce a new contract ERC20TokenManager that inherits from ERC20 and implements AccessControl 
- Define MINTER_ROLE and grant it to the contract deployer
- Implement mint and burn functions to handle token creation and destruction Use OpenZeppelin contracts as dependencies (ERC20, AccessControl)

- Create an ERC20TokenManager contract that inherits from ERC20 and AccessControl 
- Implement mint and burn functions with access control for MINTER_ROLE

- Add new AdminVault contract with role-based access control and reentrancy protection 
- Implement vault initialization with ERC20 token support 
- Add functions for token transfers between admin and users 
- Include safety features like pause mechanism and token recovery 
- Add comprehensive event logging for all major operations

### CODE REVIEW

File: src/interfaces/erc20TokenManager.sol [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+
+contract ERC20TokenManager is ERC20, AccessControl {
```
Comment (halink0803) [inline src/interfaces/erc20TokenManager.sol:7]: this file should be named like IERC20TokenManager? and it should be only interface?

File: src/Counter.sol [hunk 2]
```diff
@@ -1,14 +1,30 @@
 // SPDX-License-Identifier: UNLICENSED
 pragma solidity ^0.8.13;
 
-contract Counter {
-    uint256 public number;
+library Counters {
```
Comment (halink0803) [inline src/Counter.sol:4]: The counter is just the sample contract from foundry, or we will use the counter contract for any purpose on the project?

File: src/core/adminVault.sol [hunk 3]
```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/utils/Pausable.sol";
+
+contract superAdminVault is AccessControl, ReentrancyGuard, Pausable {
```
Comment (halink0803) [inline src/core/adminVault.sol:9]: file name is `adminVault.sol` but the contract is `superAdminVault.sol`, should changes the contract name or the file name for consistency.

File: src/core/adminVault.sol [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        require(token.transfer(admin, amount), "Transfer to admin failed");
+
+        emit TokensTransferredToAdmin(admin, amount);
+    }
+
+    function transferToUser(
+        address admin,
+        address user,
+        uint256 amount
+    ) external onlyRole(MEMBERSHIP_CONTRACT_ROLE) nonReentrant whenNotPaused {
```
Comment (halink0803) [inline src/core/adminVault.sol:87]: Is `MEMBERSHIP_CONTRACT_ROLE` assign to one or many user? If many then any of them can call this contract and transfer the token out?

**Review by halink0803 — CHANGES_REQUESTED** [review]
I think the ERC20TokenManager is only the normal ERC20 with mint and burn function, which is quite normal to a ERC20 token. I think it should be named in more relatable way, such as the token name or the token stand for (MembershipToken, GovernanceToken, etc), instead of `TokenManager`.

File: src/interfaces/erc20TokenManager.sol (hunk 1 above)
Comment (olasunkanmi-SE) [reply src/interfaces/erc20TokenManager.sol:7]: True. It will be renamed

File: src/Counter.sol (hunk 2 above)
Comment (olasunkanmi-SE) [reply src/Counter.sol:4]: Yes, we will use the counter contract for increment and decrement of mappings within the application

File: src/core/adminVault.sol (hunk 4 above)
Comment (olasunkanmi-SE) [reply src/core/adminVault.sol:87]: this is only assigned to the admin. MEMBERSHIP_CONTRACT_ROLE

**Pull request closed without merging**